File: app/src/main/java/email/schaal/ocreader/database/model/Item.java
/*
 * Copyright (C) 2015 Daniel Schaal <daniel@schaal.email>
 *
 * This file is part of OCReader.
 *
 * OCReader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * OCReader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OCReader.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package email.schaal.ocreader.database.model;

import android.support.annotation.Nullable;
import android.util.Log;

import java.util.Date;

import email.schaal.ocreader.database.Queries;
import io.realm.Realm;
import io.realm.RealmObject;
import io.realm.annotations.Index;
import io.realm.annotations.PrimaryKey;

/**
 * RealmObject representing a feed Item.
 */
public class Item extends RealmObject implements Insertable {
    private static final String TAG = Item.class.getName();

    @PrimaryKey
    private long id;
    public final static String ID = "id";

    private String guid;

    private String guidHash;
    public final static String GUID_HASH = "guidHash";

    @Nullable
    private String url;

    private String title;
    public final static String TITLE = "title";

    private String author;

    private Date pubDate;
    public static final String PUB_DATE = "pubDate";

    private Date updatedAt;
    public static final String UPDATED_AT = "updatedAt";

    private String body;
    public static final String BODY = "body";

    private String enclosureMime;
    private String enclosureLink;

    private Feed feed;
    public final static String FEED = "feed";

    private long feedId;
    public final static String FEED_ID = "feedId";

    private boolean unread;
    public final static String UNREAD = "unread";

    private boolean unreadChanged = false;
    public final static String UNREAD_CHANGED = "unreadChanged";

    private boolean starred;
    public static final String STARRED = "starred";

    private boolean starredChanged = false;
    public final static String STARRED_CHANGED = "starredChanged";

    private long lastModified;
    public static final String LAST_MODIFIED = "lastModified";

    /** @since 8.4.0 **/
    @Index
    private String fingerprint;
    public static final String FINGERPRINT = "fingerprint";

    @Index
    private String contentHash;
    public static final String CONTENT_HASH = "contentHash";

    public Item() {
    }

    public Item(long id) {
        this.id = id;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getGuid() {
        return guid;
    }

    public void setGuid(String guid) {
        this.guid = guid;
    }

    public String getGuidHash() {
        return guidHash;
    }

    public void setGuidHash(String guidHash) {
        this.guidHash = guidHash;
    }

    @Nullable
    public String getUrl() {
        return url;
    }

    public void setUrl(@Nullable String url) {
        this.url = url;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public Date getPubDate() {
        return pubDate;
    }

    public void setPubDate(Date pubDate) {
        this.pubDate = pubDate;
    }

    public Date getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Date updatedAt) {
        this.updatedAt = updatedAt;
    }

    public String getBody() {
        return body;
    }

    public void setBody(String body) {
        this.body = body;
    }

    public String getEnclosureMime() {
        return enclosureMime;
    }

    public void setEnclosureMime(String enclosureMime) {
        this.enclosureMime = enclosureMime;
    }

    public String getEnclosureLink() {
        return enclosureLink;
    }

    public void setEnclosureLink(String enclosureLink) {
        this.enclosureLink = enclosureLink;
    }

    public long getFeedId() {
        return feedId;
    }

    public void setFeedId(long feedId) {
        this.feedId = feedId;
    }

    public Feed getFeed() {
        return feed;
    }

    public void setFeed(Feed feed) {
        this.feed = feed;
    }

    public boolean isUnread() {
        return unread;
    }

    public void setUnread(boolean unread) {
        if(isValid() && this.unread != unread) {
            unreadChanged = !unreadChanged;
            feed.incrementUnreadCount(unread ? 1 : -1);
        }
        this.unread = unread;
    }

    public boolean isStarred() {
        return starred;
    }

    public void setStarred(boolean starred) {
        if(isValid() && this.starred != starred) {
            starredChanged = !starredChanged;
            feed.incrementStarredCount(starred ? 1 : -1);
        }
        this.starred = starred;
    }

    public boolean isUnreadChanged() {
        return unreadChanged;
    }

    public void setUnreadChanged(boolean unreadChanged) {
        this.unreadChanged = unreadChanged;
    }

    public boolean isStarredChanged() {
        return starredChanged;
    }

    public void setStarredChanged(boolean starredChanged) {
        this.starredChanged = starredChanged;
    }

    public long getLastModified() {
        return lastModified;
    }

    public void setLastModified(long lastModified) {
        this.lastModified = lastModified;
    }

    public String getFingerprint() {
        return fingerprint;
    }

    public String getContentHash() {
        return contentHash;
    }

    public void setContentHash(String contentHash) {
        this.contentHash = contentHash;
    }

    public void setFingerprint(String fingerprint) {
        this.fingerprint = fingerprint;
    }

    @Override
    public void insert(Realm realm) {
        if (getTitle() == null) {
            // Reduced item
            final Item fullItem = realm.where(Item.class).equalTo(Item.CONTENT_HASH, getContentHash()).findFirst();
            if (fullItem != null) {
                fullItem.setUnread(isUnread());
                fullItem.setStarred(isStarred());
            } else {
                Log.w(TAG, "Full item is not available");
            }
        } else {
            // new full item
            setFeed(Feed.getOrCreate(realm, getFeedId()));
            realm.insertOrUpdate(this);
        }

    }

    @Override
    public void delete(Realm realm) {
        deleteFromRealm();
    }
}

File: app/src/main/java/email/schaal/ocreader/database/model/Item.kt
/*
 * Copyright © 2020. Daniel Schaal <daniel@schaal.email>
 *
 * This file is part of ocreader.
 *
 * ocreader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ocreader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 */

package email.schaal.ocreader.database.model

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Parcelable
import io.realm.Realm
import io.realm.RealmModel
import io.realm.RealmObject
import io.realm.Sort
import io.realm.annotations.PrimaryKey
import io.realm.annotations.RealmClass
import io.realm.annotations.RealmField
import io.realm.kotlin.isManaged
import io.realm.kotlin.where
import kotlinx.android.parcel.Parcelize
import java.util.*

@Parcelize
@RealmClass
open class Item(
        @PrimaryKey var id: Long = 0,
        var guid: String? = null,
        var guidHash: String? = null,
        var url: String? = null,
        var title: String? = null,
        var author: String? = null,
        var pubDate: Date? = null,
        var updatedAt: Date? = null,
        var body: String = "",
        var enclosureMime: String? = null,
        var enclosureLink: String? = null,
        var feed: Feed? = null,
        var feedId: Long = 0,
        @RealmField(name = "unread")
        private var actualUnread: Boolean = true,
        var unreadChanged: Boolean = false,
        @RealmField(name = "starred")
        private var actualStarred: Boolean = false,
        var starredChanged: Boolean = false,
        var lastModified: Long = 0,
        var fingerprint: String? = null,
        var contentHash: String? = null,
        var active: Boolean = true
) : RealmModel, Insertable, Parcelable {
    companion object {
        const val UNREAD = "actualUnread"
        const val STARRED = "actualStarred"

        fun removeExcessItems(realm: Realm, maxItems: Int) {
            val itemCount = realm.where<Item>().count()
            if (itemCount > maxItems) {
                realm.where<Item>()
                        .equalTo(UNREAD, false)
                        .equalTo(STARRED, false)
                        .equalTo(Item::active.name, false)
                        .sort(Item::lastModified.name, Sort.ASCENDING)
                        .limit(itemCount - maxItems)
                        .findAll()
                        .deleteAllFromRealm()
            }
        }

    }

    class Builder {
        var id: Long = 0
        var guid: String? = null
        var guidHash: String? = null
        var url: String? = null
        var title: String? = null
        var author: String? = null
        var pubDate: Date? = null
        var updatedAt: Date? = null
        var body: String = ""
        var enclosureMime: String? = null
        var enclosureLink: String? = null
        var feed: Feed? = null
        var feedId: Long = 0
        var unread: Boolean = true
        var unreadChanged: Boolean = false
        var starred: Boolean = false
        var starredChanged: Boolean = false
        var lastModified: Long = 0
        var fingerprint: String? = null
        var contentHash: String? = null
        var active: Boolean = true

        fun build() : Item {
            return Item(this)
        }
    }

    constructor(builder: Builder) :this(
            builder.id,
            builder.guid,
            builder.guidHash,
            builder.url,
            builder.title,
            builder.author,
            builder.pubDate,
            builder.updatedAt,
            builder.body,
            builder.enclosureMime,
            builder.enclosureLink,
            builder.feed,
            builder.feedId,
            builder.unread,
            builder.unreadChanged,
            builder.starred,
            builder.starredChanged,
            builder.lastModified,
            builder.fingerprint,
            builder.contentHash,
            builder.active
    )

    var unread: Boolean
        get() { return actualUnread }
        set(value) {
            if(isManaged() && actualUnread != value) {
                unreadChanged = !unreadChanged
                feed?.incrementUnreadCount(if(value) 1 else -1)
            }
            actualUnread = value
        }

    var starred: Boolean
        get() { return actualStarred }
        set(value) {
            if(isManaged() && actualStarred != value) {
                starredChanged = !starredChanged
                feed?.incrementStarredCount(if(value) 1 else -1)
            }
            actualStarred = value
        }

    override fun insert(realm: Realm) {
        if(title == null) {
            val fullItem = realm.where<Item>().equalTo(Item::contentHash.name, contentHash).findFirst()
            fullItem?.unread = unread
            fullItem?.starred = starred
        } else {
            feed = Feed.getOrCreate(realm, feedId)
        }
        realm.insertOrUpdate(this)
    }

    override fun delete(realm: Realm) {
        RealmObject.deleteFromRealm(this)
    }

    fun play(context: Context) {
        if(enclosureLink != null) {
            val playIntent = Intent(Intent.ACTION_VIEW)
            playIntent.data = Uri.parse(enclosureLink)
            context.startActivity(playIntent)
        }
    }
}